src/ngResource/resource.js

```diff
@@ -743,7 +746,14 @@ angular.module('ngResource', ['ng']).
               extend({}, extractParams(data, action.params || {}), params),
               action.url);
 
-            var promise = $http(httpConfig).then(function(response) {
+            // Start the promise chain
+            var promise = $q.
+              resolve(httpConfig).
+              then(requestInterceptor).
+              catch(requestErrorInterceptor).
+              then($http);
```

Comment:
Interested to know why you "reimplemented" request interceptors here rather than modifying the `httpConfig` to add the interceptors to the request `httpConfig` (or a clone of it) and rely upon `$http` to deal with the intercepting.

Reply:
Currently `$http` does not support per-request interceptors. Are you suggesting add that as a new feature to `$http` (and then use it to implement `$resource` request interceptors?

Reply:
This would actually be my preferred solution, as the resource interceptors do not work 100% like httpInterceptors anyway (see https://github.com/angular/angular.js/issues/11409. Note that even though #13273 is linked in this issue as a solution, adding request/Error interceptors does not actually solve the issue); and because http transforms do not support changing all variables (headers cannot be changed, see https://github.com/angular/angular.js/issues/12095)
So you would be able to set `interceptors` in the http config and `httpInterceptors` in ngResource.

Obviously this would be a much bigger change which would make http much more complex.

Reply:
> This would actually be my preferred solution, as the resource interceptors do not work 100% like httpInterceptors anyway

`$resource` response interceptors are special, so indeed do not work exactly as `$http` interceptors. But request interceptors work pretty much the same afaict.

> So you would be able to set interceptors in the http config and httpInterceptors in ngResource.
> Obviously this would be a much bigger change which would make http much more complex.

I agree with both observations :wink:

Reply:
If we go with requestInterceptors (fine with me), then we should take this opportunity to document responseInterceptors of resource and the differences to http behavior, preferrably in the same or a follow up PR.

Reply:
What differences do you have in mind? The only deifference that I know of is the extra `response.resource` property and this is already documented.

Reply:
I didn't realise that we couldn't do per-request interceptors in `$http`!

Reply:
@gkalpak I'll have to check, but since https://github.com/angular/angular.js/issues/11409 is still open I assumed the docs are not clear, or a workaround for this is not covered by the docs.

Reply:
Just saw there's an extra issue for the inconsistency in response intereceptors: https://github.com/angular/angular.js/issues/9334

Reply:
#9334, is about inconsistency between `response`/`responseError` interceptors in `$resource` itself (not about inconsistency against `$http` interceptors). #9334 has valid points (that are probably still worth doing something about for 1.7), but is not related to `request`/`requestError` interceptors (which are doing the "right thing"&trade; afaict).

Reply:
Yeah, it's not directly related. I just thought it could be a good time to clarify any docs in general. But it's not a must.